examples/qbf/Synthesis.hs

```diff
@@ -0,0 +1,109 @@
+-- | An example program that solves a circuit synthesis problem with the Boolean Chain approach,
+-- as cited in "Circuit Minimization with QBF-Based Exact Synthesis" by Reichl et al. 2023
+-- <https://doi.org/10.1609/aaai.v37i4.25524/>.
+
+module Main where
+
+import Prelude hiding (not, (&&), and, or, product)
+
+import Ersatz
+import Ersatz.Relation
+import Ersatz.Counting
+
+import qualified Data.Array as A
+import Data.Array (Array, Ix)
+
+import Control.Monad.State.Lazy (StateT)
+import Control.Monad (guard)
+
+import Text.Printf (printf)
+import Data.List (sortOn)
+
+import qualified Data.ByteString.Lazy as B
+import qualified Data.ByteString.Lazy.Char8 as C
+import Data.Functor.Identity
+
+
+-- | AIG where @1 && 2@ is equivalent to @4@
+main :: IO ()
+main = do
+  let n = 2 -- number of inputs
+      l = 2 -- number of steps (gates)
+      m = 1 -- number of outputs
+      f :: [Bit] -> [Bit]
+      f xs = [and xs]
+  formulaSize $ problem f n l m
+  solve $ problem f n l m
+
+-- | AIG where @atmost 2 [1,2,3,4]@ is equivalent to @10@
+aig_atmost :: IO ()
+aig_atmost = do
+  let n = 4 -- number of inputs
+      l = 6 -- number of steps (gates)
+      m = 1 -- number of outputs
+      f :: [Bit] -> [Bit]
+      f xs = [atmost 2 xs]
+  formulaSize $ problem f n l m
+  solve $ problem f n l m
+
+-- | AIG where @exactly 2 [1,2,3]@ is equivalent to @8@
+aig_exactly :: IO ()
+aig_exactly = do
+  let n = 3 -- number of inputs
+      l = 5 -- number of steps (gates)
+      m = 1 -- number of outputs
+      f :: [Bit] -> [Bit]
+      f xs = [exactly 2 xs]
+  formulaSize $ problem f n l m
+  solve $ problem f n l m

```

Comment:
Neither `aig_atmost` nor `aig_exactly` are used at runtime. Should they be?

Reply:
I used one as the main function and removed the other, but mentioned the corresponding problem in the comments.